src/bin/busd.rs

```diff
@@ -26,6 +28,48 @@ struct Args {
     #[clap(long)]
     #[arg(value_enum, default_value_t = AuthMechanism::External)]
     auth_mechanism: AuthMechanism,
+
+    /// File descriptor to which readiness notifications are sent.
+    ///
+    /// Once the server is listening to connections on the specified socket, it will print
+    /// `READY=1\n` into this file descriptor and close it.
+    ///
+    /// This readiness notification mechanism which works on both systemd and s6.
+    #[clap(long)]
+    ready_fd: Option<i32>,
+}
+
+impl Args {
+    /// Returns a `ReadyFd` for the provided file descriptor
+    ///
+    /// The inner `ready_fd` value is consumed to avoid creating multiple instances of the same
+    /// file descriptor, which would result in undefined behaviour.
+    ///
+    /// # Safety
+    ///
+    /// The `ready_fd` argument must be a valid and open file descriptor
+    fn take_ready_fd(&mut self) -> Option<ReadyFd> {
+        // SAFETY: no other function can access `ready_fd` because it is being taken here, so there
+        // is only one owner for this file descriptor.
+        self.ready_fd
+            .take()
+            .map(|fd| unsafe { ReadyFd(File::from_raw_fd(fd)) })
+    }
+}
+
+struct ReadyFd(File);
+
+impl ReadyFd {
+    /// Notify parent that the service is ready to accept connections.
+    ///
+    /// Consumes the `ReadyFd`, which will then drops (and closes) the file descriptor.
+    ///
+    /// # Errors
+    ///
+    /// Returns any error as described in [`std::io::Write::write_all`].
+    fn notify_ready(mut self) -> std::io::Result<()> {
+        self.0.write_all(b"READY=1\n")
+    }
```

Comment:
Don't think these types are needed. We only need to call `File::from_raw_fd(fd).write_all(..)`. We don't need to care about fd lifecycle or document that then cause we know it's just the main function that will use this once only. No need to even make `args` mutable.

Reply:
I think I over-engineered this expecting to pass `ReadyFd` into a few nested functions and then realised it was use right here in `main`.